addons/web/controllers/doc.py

```diff
@@ -0,0 +1,496 @@
+    ]
+    default: typing.Any | inspect._empty
+    annotation: str | None
+    doc: str | None
+
+    @classmethod
+    def from_inspect(cls, parameter):
+        return cls(
+            name=parameter.name,
+            kind=parameter.kind.name,
+            default=parameter.default,
+            annotation=stringify_annotation(parameter.annotation),
+            doc=None,
+        )
+
+    def as_dict(self):
+        d = vars(self)
+        if self.kind == 'POSITIONAL_OR_KEYWORD':
+            # most (99%) params are POSITIONAL_OR_KEYWORD
+            # make the export smaller by ignoring those
+            d.pop('kind')
+        if self.annotation is None:
+            d.pop('annotation')
+        if self.doc is None:
+            d.pop('doc')
+        if self.default is inspect._empty:
+            d.pop('default')
+        else:
+            # ignore the default value when it is not json serializable
+            try:
+                json.dumps(self.default)
+            except ValueError:
+                d.pop('default')
+        return d
+
+
+@dataclasses.dataclass
+class Return:
+    annotation: str | None
+    doc: str | None
+
+    @classmethod
+    def from_inspect(cls, return_annotation):
+        return cls(stringify_annotation(return_annotation), doc=None)
+
+    def as_dict(self):
+        d = vars(self)
+        if self.annotation is None:
+            d.pop('annotation')
+        if self.doc is None:
+            d.pop('doc')
+        return d
+
+
+# This class could had been a python module, but lazy_classproperty
+# works much better than odoo.tools.lazy.
+class _DocUtils:
+    """ Helpers for docutils """
+    @lazy_classproperty
+    def _new_docutils_root(cls):
+        # surely there's a better way, but that'll do
+        return docutils.core.publish_doctree("").copy
+
+    @classmethod
+    def _make_settings(cls, writer_name, settings_overrides):
+        pub = docutils.core.Publisher()
+        pub.set_components('standalone', 'restructuredtext', writer_name)
+        pub.process_programmatic_settings(None, settings_overrides, None)
+        return pub.settings
+
+    @lazy_classproperty
+    def _settings_pseudoxml(cls):
+        return cls._make_settings('pseudoxml', {
+            'report_level': 3,
+            'halt_level': 5,
+        })
+
+    @lazy_classproperty
+    def _settings_html(cls):
+        return cls._make_settings('html', {
+            'report_level': 3,
+            'halt_level': 5,
+            'embed_stylesheet': False,
+        })
+
+    @classmethod
+    def tree(cls, docstring):
+        with contextlib.redirect_stderr(io.StringIO()) as stderr:
+            doctree = docutils.core.publish_doctree(
+                docstring,
+                settings=cls._settings_pseudoxml,
+            )
+            if stderr.tell():
+                logger.warning(RST_PARSE_ERROR.format(docstring, stderr.getvalue()))
+            return doctree
+
+    @classmethod
+    def html(cls, tree):
```

Comment:
I wonder if we shouldn't sanitize the html it returns

@std-odoo wanna try and find an XSS using docstrings? :D

Reply:
Docutils escape the html :)
```py
"""
<script>alert("haha")</script>
"""
```
=>
```html
<div class="document">\n\n\n&lt;script&gt;alert(&quot;haha&quot;)&lt;/script&gt;\n</div>
```